### PR TITLE
fix(manager/pub): extract sdk constraint correctly

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

Fix the pub manager to extract SDK constraint correctly for running artifacts update by parsing the `pubspec.lock` file which is a YAML file instead of using regex.

## Context

I don't think there are any logs when extracting the SDK constraint, but basically the current implementation of `getFlutterConstraint` and `getDartConstraint` always return `undefined` for my Flutter app repository, which means the latest version of Flutter is always being used.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/pub/types.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export interface PubspecSdk {
```
Comment (viceice) [inline lib/modules/manager/pub/types.ts:1]: use zod schema validation, check package.json inside npm manager for sample

File: lib/modules/manager/pub/artifacts.ts [hunk 2]
```diff
@@ -9,18 +10,23 @@ import {
...
-
-function getDartConstraint(lockFileContent: string): string | undefined {
-  return regEx(/^\tdart: ['"](?<dartVersion>.*)['"]$/m).exec(lockFileContent)
-    ?.groups?.dartVersion;
+export function getConstraint(
+  lockFileContent: string,
+  lockFileSdkKey: keyof PubspecSdk
+): string | undefined {
+  try {
+    const data = load(lockFileContent, { json: true }) as PubspecLock;
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:21]: parse once with lazy, checkout npm manager for sample

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/pub/artifacts.ts (hunk 2 above)
Comment (zeshuaro) [reply lib/modules/manager/pub/artifacts.ts:21]: Sorry, not sure what you meant with this and couldn't find the sample, would you be able to share the link to the sample?

Comment (viceice) [reply lib/modules/manager/pub/artifacts.ts:21]: ups, PR wasn't merged yet

https://github.com/renovatebot/renovate/pull/23131

File: lib/modules/manager/pub/utils.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import { PubspecLockSchema } from './schema';
+
+export function lazyParsePubspeckLock(
+  fileContent: string
+): Lazy<PubspecLockSchema | undefined> {
+  return new Lazy(() => parsePubspecLock(fileContent));
+}
+
+function parsePubspecLock(fileContent: string): PubspecLockSchema | undefined {
+  const res = PubspecLockSchema.safeParse(fileContent);
```
Comment (zeshuaro) [inline lib/modules/manager/pub/utils.ts:11]: I updated the PR to use zod as suggested, but I'm having trouble with it here. Basically, I'm not sure how I should convert the file content string into an object for zod to validate.

File: lib/modules/manager/pub/artifacts.ts [hunk 4]
```diff
@@ -63,9 +52,11 @@ export async function updateArtifacts({
       );
     }
 
+    const oldLazyPubspecLock = lazyParsePubspeckLock(oldLockFileContent);
+    const oldPubspecLock = oldLazyPubspecLock.getValue();
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:56]: inline this, so the lockfile is only read when no constraints are manually configured. that's the idea behind the lazy.

File: lib/modules/manager/pub/utils.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/manager/pub/utils.ts:11]: use yaml-js to parse yaml to object

File: lib/modules/manager/pub/__fixtures__/pubspec.invalid.lock [hunk 5]
```diff
@@ -0,0 +1,2 @@
+clearly:
```
Comment (viceice) [inline lib/modules/manager/pub/__fixtures__/pubspec.invalid.lock:1]: can you please inline this small fixture?

File: lib/modules/manager/pub/utils.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import { Lazy } from '../../../util/lazy';
+import { PubspecLockSchema } from './schema';
+
+export function lazyParsePubspeckLock(
+  fileContent: string
+): Lazy<PubspecLockSchema | undefined> {
+  return new Lazy(() => parsePubspecLock(fileContent));
+}
+
+function parsePubspecLock(fileContent: string): PubspecLockSchema | undefined {
```
Comment (viceice) [inline lib/modules/manager/pub/utils.ts:11]:
```suggestion
function parsePubspecLock(fileContent: string): PubspecLockSchema | null {
```

File: lib/modules/manager/pub/utils.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  try {
+    const data = load(fileContent, { json: true });
+    const res = PubspecLockSchema.safeParse(data);
+    if (res.success) {
+      return res.data;
+    }
+  } catch {
+    // Do nothing
+  }
```
Comment (viceice) [inline lib/modules/manager/pub/utils.ts:20]:
```suggestion
  }
  return null;
```

**Review by viceice — COMMENTED** [review]
I prefer null over undefined

File: lib/modules/manager/pub/utils.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+function parsePubspecLock(fileContent: string): PubspecLockSchema | null {
+  try {
+    const data = load(fileContent, { json: true });
+    const res = PubspecLockSchema.safeParse(data);
+    if (res.success) {
+      return res.data;
+    } else {
+      logger.debug(res.error, 'Error parsing pubspec.lock file');
```
Comment (viceice) [inline lib/modules/manager/pub/utils.ts:19]: please pass the lockfile name from caller, maybe someone has multiple lockfiles inside different folders.

File: lib/modules/manager/pub/schema.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { z } from 'zod';
+import { Yaml } from '../../../util/schema-utils';
+
+export const PubspecLockSchema = z.object({
+  packages: z.record(z.any()),
```
Comment (zharinov) [inline lib/modules/manager/pub/schema.ts:5]: Let's omit this field completely if it isn't being used anywhere in Renovate

File: lib/modules/manager/pub/artifacts.spec.ts [hunk 10]
```diff
@@ -17,8 +18,8 @@ jest.mock('../../datasource');
 process.env.CONTAINERBASE = 'true';
 
 const lockFile = 'pubspec.lock';
-const oldLockFileContent = 'Old pubspec.lock';
-const newLockFileContent = 'New pubspec.lock';
+const oldLockFileContent = Fixtures.get('pubspec.1.lock');
+const newLockFileContent = Fixtures.get('pubspec.2.lock');
```
Comment (zharinov) [inline lib/modules/manager/pub/artifacts.spec.ts:22]: Please, let's use `codeBlock` instead (see other usages in the codebase)

File: lib/modules/manager/pub/artifacts.ts [hunk 11]
```diff
@@ -63,9 +52,14 @@ export async function updateArtifacts({
...
+    const oldLazyPubspecLock = lazyParsePubspeckLock(
+      lockFileName,
+      oldLockFileContent
+    );
     const constraint = isFlutter
-      ? config.constraints?.flutter ?? getFlutterConstraint(oldLockFileContent)
-      : config.constraints?.dart ?? getDartConstraint(oldLockFileContent);
+      ? config.constraints?.flutter ??
+        oldLazyPubspecLock.getValue()?.sdks.flutter
+      : config.constraints?.dart ?? oldLazyPubspecLock.getValue()?.sdks.dart;
```
Comment (zharinov) [inline lib/modules/manager/pub/artifacts.ts:62]: It reads a bit hard to me. What about something like this:

```ts
const sdk = isFlutter ? 'flutter' : 'dart';
let constraint = config.constraints?.[sdk];
if (!constraint) {
  // Lazy part is not necessary now
  constraint = ...
}
```

**Review by zharinov — CHANGES_REQUESTED** [review]

File: lib/modules/manager/pub/artifacts.ts (hunk 11 above)
Comment (viceice) [reply lib/modules/manager/pub/artifacts.ts:62]: sounds good to me

File: lib/modules/manager/pub/schema.ts (hunk 9 above)
Comment (viceice) [reply lib/modules/manager/pub/schema.ts:5]:
```suggestion
```

File: lib/modules/manager/pub/utils.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+function parsePubspecLock(
+  fileName: string,
+  fileContent: string
+): PubspecLockSchema | null {
+  const res = PubspecLockYaml.safeParse(fileContent);
+  if (res.success) {
+    return res.data;
+  } else {
+    logger.debug(res.error, `Error parsing ${fileName} file`);
```
Comment (viceice) [inline lib/modules/manager/pub/utils.ts:20]:
```suggestion
    logger.debug({ err: res.error, fileName }, `Error parsing pubspec lockfile.`);
```

File: lib/modules/manager/pub/artifacts.spec.ts (hunk 10 above)
Comment (zharinov) [reply lib/modules/manager/pub/artifacts.spec.ts:22]: To make code blocks smaller, I believe we can remove `packages` section

File: lib/modules/manager/pub/artifacts.ts [hunk 13]
```diff
@@ -63,9 +52,15 @@ export async function updateArtifacts({
...
-    const constraint = isFlutter
-      ? config.constraints?.flutter ?? getFlutterConstraint(oldLockFileContent)
-      : config.constraints?.dart ?? getDartConstraint(oldLockFileContent);
+    let constraint = config.constraints?.[toolName];
+    if (!constraint) {
+      const pubspecLock = parsePubspecLock(lockFileName, oldLockFileContent);
+      const pubspecLockSdks = pubspecLock?.sdks;
+      if (pubspecLockSdks) {
+        constraint = isFlutter ? pubspecLockSdks.flutter : pubspecLockSdks.dart;
+      }
```
Comment (zharinov) [inline lib/modules/manager/pub/artifacts.ts:61]:
```suggestion
      constraint = pubspecLock?.sdks[toolName];
```

**Review by zharinov — APPROVED** [review]

**Review by zharinov — APPROVED** [review]

File: lib/modules/manager/pub/utils.spec.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import { codeBlock } from 'common-tags';
+import { parsePubspecLock } from './utils';
+
+describe('modules/manager/pub/utils', () => {
+  describe('lazyParsePubspeckLock', () => {
```
Comment (viceice) [inline lib/modules/manager/pub/utils.spec.ts:5]:
```suggestion
  describe('parsePubspeckLock', () => {
```

**Review by viceice — COMMENTED** [review]
just a small z

**Review by zharinov — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**